src/main/java/org/reactome/addlinks/dataretrieval/pharos/PharosDataRetriever.java

```diff
@@ -0,0 +1,159 @@
+				String updatedQuery = this.getQuery()
+										.replace(PharosDataRetriever.TOP_AMOUNT_TOKEN, Integer.toString(PharosDataRetriever.batchSize))
+										.replace(PharosDataRetriever.SKIP_AMOUNT_TOKEN, Integer.toString(numRequests * PharosDataRetriever.batchSize));
+				StringEntity stringEntity = new StringEntity(updatedQuery);
+				stringEntity.setContentType(PharosDataRetriever.APPLICATION_JSON);
+				post.setEntity(stringEntity);
+				post.setHeader(HttpHeaders.ACCEPT, PharosDataRetriever.APPLICATION_JSON);
+				post.setHeader(HttpHeaders.CONTENT_TYPE, PharosDataRetriever.APPLICATION_JSON);
+				int retries = this.numRetries;
+				boolean done = retries + 1 <= 0;
+				while (!done)
+				{
+					try (CloseableHttpClient client = HttpClients.createDefault();
+						CloseableHttpResponse response = client.execute(post, context); )
+					{
+						int statusCode = response.getStatusLine().getStatusCode();
+						// If status code was not 200, we should print something so that the users know that an unexpected response was received.
+						if (statusCode != HttpStatus.SC_OK)
+						{
+							if (String.valueOf(statusCode).startsWith("4") || String.valueOf(statusCode).startsWith("5"))
+							{
+								logger.error("Response code was 4xx/5xx: {}, Status line is: {}", statusCode, response.getStatusLine());
+							}
+							else
+							{
+								logger.warn("Response was not \"200\". It was: {}", response.getStatusLine());
+							}
+						}
+						else
+						{
+							// result could be flattened with jq using the jq expression: ".data.targets.targets[].uniprot"
+							String jsonResponse = EntityUtils.toString(response.getEntity());
+							JSONObject jsonObj = new JSONObject(jsonResponse);
+
+							int numProcessed = processJSONArray(fw, jsonObj);
+							moreRecords = numProcessed >= 1;
+							logger.info("Completed request #{}", numRequests);
+							numRequests++;
+						}
+						done = true;
+					}
+					catch (ConnectTimeoutException e)
+					{
+						// we will only be retrying the connection timeouts, defined as the time required to establish a connection.
+						// we will not handle socket timeouts (inactivity that occurs after the connection has been established).
+						// we will not handle connection manager timeouts (time waiting for connection manager or connection pool).
+						e.printStackTrace();
+						logger.info("Failed due to ConnectTimeout, but will retry {} more time(s).", retries);
+						retries--;
+						done = retries + 1 <= 0;
+						if (done)
+						{
+							throw new Exception("Connection timed out. Number of retries ("+this.numRetries+") exceeded. No further attempts will be made.", e);
+						}
+					}
+					catch (HttpHostConnectException e)
+					{
+						logger.error("Could not connect to host {} !",post.getURI().getHost());
+						e.printStackTrace();
+						throw e;
+					}
+					catch (IOException e) {
+						logger.error("Exception caught: {}",e.getMessage());
+						throw e;
+					}
```

Comment:
Could the stack trace print on line 132 be part of the logger message on line 131?

`logger.error("Could not connect to host {}: {} !", post.getURI().getHost(), e.getMessage())`;

This is done on line 136.  It should be okay not to line 132 because the exception gets rethrown.

Reply:
Just to clarify: your code suggestion will _not_ print the stack trace, it will _only_ print the message. I'm fine with that, if that's what you meant.

Reply:
Yeah, I was thinking to print the message to be in sync with the way the next exception block logs the exception.  You could pass the entire exception object to the logger as the last argument instead of the message if you want so the entire stack trace is logged.

I also thought since the exception is rethrown another part of the code may print the stack trace so it may be okay to take out the stack trace to screen here.

Reply:
The statement from the "next exception block" (`logger.error("Exception caught: {}",e.getMessage());`) does not log the entire exception. Only the message.

I changed the code to only log the messages, I thought that's what you wanted.

Reply:
I'm okay either way - I just was mentioning another possibility.  I'm fine with the way the code is now.